src/Core/ConstraintValidator/Constraints/ValidReductionType.php

```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * 2007-2020 PrestaShop SA and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2020 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints;
+
+use PrestaShop\PrestaShop\Core\ConstraintValidator\ValidReductionTypeValidator;
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Constraint for validating reduction type
+ */
+final class ValidReductionType extends Constraint
+{
+    public $message = 'Reduction type "%type%" is invalid. Allowed types are: %types%.';
```

Comment:
This one is really used? if yes, can be a const :thinking:

Reply:
I think this kind of validation is needed because even if we use form when it's possible, changes can be done by CQRS commands for example. But I also think @matks or @jolelievre are better than me to answer you correctly (they're our CQRS experts!)
(really sorry for the delay, little bit busy :sweat_smile: )

Reply:
I think it would be nice to have this ValidReductionType class. It could be a helper class for developers who want to use it and easily strengthen the user experience of their custom form.

However if we can reuse `Choice` logic it's better indeed. So maybe solution is a simple children class:
```
<?php

class ValidReductionType extends Choice
{
    public $choices = ['percentage', 'amount', 'free_shipping'];

}
```

Something like that ?

Another nice point of having a dedicated Constraint class is that @LouiseBonnard can customize the error message we want to display to user.